demo/site/src/documentTypes/Page.tsx

```diff
@@ -41,18 +42,77 @@ const pageQuery = gql`
     ${topMenuPageTreeNodeFragment}
 `;
 
-export async function Page({ pageTreeNodeId, scope }: { pageTreeNodeId: string; scope: GQLPageTreeNodeScopeInput }) {
+type Props = { pageTreeNodeId: string; scope: GQLPageTreeNodeScopeInput };
+
+async function fetchData({ pageTreeNodeId, scope }: Props) {
     const { previewData } = previewParams() || { previewData: undefined };
     const graphQLFetch = createGraphQLFetch(previewData);
 
-    const data = await graphQLFetch<GQLPageQuery, GQLPageQueryVariables>(pageQuery, {
-        pageTreeNodeId,
-        domain: scope.domain,
-        language: scope.language,
-    });
+    const props = await graphQLFetch<GQLPageQuery, GQLPageQueryVariables>(
+        pageQuery,
+        {
+            pageTreeNodeId,
+            domain: scope.domain,
+            language: scope.language,
+        },
+        { method: "GET" }, //for request memoization
+    );
+
+    if (!props.pageContent) throw new Error("Could not load page content");
+    const document = props.pageContent.document;
+    if (!document) {
+        return null;
+    }
+    if (document.__typename != "Page") throw new Error(`invalid document type, expected Page, got ${document.__typename}`);
+
+    return {
+        ...props,
+        pageContent: {
+            ...props.pageContent,
+            document,
+        },
+    };
+}
+
+export async function generateMetadata({ pageTreeNodeId, scope }: Props, parent: ResolvingMetadata): Promise<Metadata> {
+    const data = await fetchData({ pageTreeNodeId, scope });
+    const document = data?.pageContent?.document;
+    if (!document) {
+        return {};
+    }
+    // TODO move into library
+    return {
+        title: document.seo.htmlTitle || data.pageContent.name,
+        description: document.seo.metaDescription,
+        openGraph: {
+            title: document.seo.openGraphTitle,
+            description: document.seo.openGraphDescription,
+            type: "website",
+            url: document.seo.canonicalUrl,
+            images: document.seo.openGraphImage.block?.urlTemplate
+                ? generateImageUrl({ src: document.seo.openGraphImage.block?.urlTemplate, width: 1024 }, 1 / 1)
+                : undefined,
+        },
+        robots: {
+            index: !document.seo.noIndex,
+        },
+        alternates: {
+            canonical: document.seo.canonicalUrl,
```

Comment:
As far as i know, if there is no canonicalUrl set in the CMS we should reference to the own page.

Every time we didn't have it, SEO agencies told us that it is important.

<img width="732" alt="Screenshot 2024-05-02 at 10 22 08" src="https://github.com/vivid-planet/comet/assets/44967610/809f017e-1958-4145-9ee4-ff1f1e6cdf0c">

Reply:
I added it, although we don't have the site url env anymore - that will have to wait for the new site config